mcs/class/Mono.Security/Mono.Security.Protocol.Tls/SslClientStream.cs

```diff
@@ -588,14 +590,28 @@ private void NegotiateAsyncWorker (IAsyncResult result)
 			}
 			catch (TlsException ex)
 			{
-				// FIXME: should the send alert also be done asynchronously here and below?
-				this.protocol.SendAlert(ex.Alert);
-				negotiate.SetComplete (new IOException("The authentication or decryption has failed.", ex));
+				alert = ex.Alert;
+				exception = ex;
 			}
 			catch (Exception ex)
 			{
-				this.protocol.SendAlert(AlertDescription.InternalError);
-				negotiate.SetComplete (new IOException("The authentication or decryption has failed.", ex));
+				alert = new Alert(AlertDescription.InternalError);
+				exception = ex;
+			}
+
+			if (exception != null)
+			{
+				try
+				{
+					// FIXME: should the send alert also be done asynchronously here and below?
+					this.protocol.SendAlert(alert);
+				}
+				catch
+				{
+					// TODO: can't create AggregateException here
```

Comment:
Yes, it should probably have said "complete with the original exception" rather than "throw the original exception", but I was just keeping the original commit message that @martinpotter submitted.